File: packages/dotnet/src/AxonPush.Cli/Git.cs
using System.Diagnostics;

namespace AxonPush.Cli;

/// <summary>
/// Best-effort repository lineage, so a gate decision can name a commit.
/// Deliberately forgiving: a checkout without git is still a valid place to
/// run an evaluation.
/// </summary>
internal static class Git
{
    public static GitLineage Capture()
    {
        var commit = Run("rev-parse HEAD");
        if (string.IsNullOrEmpty(commit))
        {
            return new GitLineage();
        }

        var branch = Run("rev-parse --abbrev-ref HEAD");
        return new GitLineage
        {
            GitCommit = commit,
            GitBranch = branch is null or "HEAD" ? null : branch,
            GitDirty = !string.IsNullOrEmpty(Run("status --porcelain --untracked-files=normal")),
        };
    }

    private static string? Run(string arguments)
    {
        try
        {
            using var process = Process.Start(new ProcessStartInfo("git", arguments)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
            });
            if (process is null)
            {
                return null;
            }

            var output = process.StandardOutput.ReadToEnd().Trim();
            if (!process.WaitForExit(10_000))
            {
                process.Kill(entireProcessTree: true);
                return null;
            }

            return process.ExitCode == 0 && output.Length > 0 ? output : null;
        }
        catch (Exception exception) when (exception is not OutOfMemoryException)
        {
            return null;
        }
    }
}

File: packages/python/src/axonpush/eval/git.py
"""Best-effort repository lineage, so a gate decision can name a commit.

Deliberately forgiving: a checkout without git, or without a commit, is still a
valid place to run an evaluation.
"""

from __future__ import annotations

import subprocess


def _git(args: list[str], cwd: str | None = None) -> str | None:
    try:
        completed = subprocess.run(  # noqa: S603 - fixed argv, no shell
            ["git", *args],
            cwd=cwd,
            capture_output=True,
            text=True,
            check=False,
            timeout=10,
        )
    except (OSError, subprocess.TimeoutExpired):
        return None
    if completed.returncode != 0:
        return None
    return completed.stdout.strip() or None


def capture_git_lineage(cwd: str | None = None) -> dict[str, object]:
    """Return ``gitCommit``/``gitBranch``/``gitDirty``, or nothing outside a repo."""
    commit = _git(["rev-parse", "HEAD"], cwd)
    if not commit:
        return {}
    branch = _git(["rev-parse", "--abbrev-ref", "HEAD"], cwd)
    status = _git(["status", "--porcelain", "--untracked-files=normal"], cwd)
    lineage: dict[str, object] = {"gitCommit": commit, "gitDirty": bool(status)}
    if branch and branch != "HEAD":
        lineage["gitBranch"] = branch
    return lineage
